File: remote/src/main/java/com/kelsos/mbrc/presenters/MainViewPresenterImpl.java
package com.kelsos.mbrc.presenters;

import com.google.inject.Inject;
import com.kelsos.mbrc.annotations.PlayerAction;
import com.kelsos.mbrc.annotations.PlayerState;
import com.kelsos.mbrc.annotations.Repeat;
import com.kelsos.mbrc.annotations.Shuffle;
import com.kelsos.mbrc.domain.TrackPosition;
import com.kelsos.mbrc.events.ui.CoverChangedEvent;
import com.kelsos.mbrc.events.ui.MuteChangeEvent;
import com.kelsos.mbrc.events.ui.PlayStateChange;
import com.kelsos.mbrc.events.ui.RepeatChange;
import com.kelsos.mbrc.events.ui.TrackInfoChangeEvent;
import com.kelsos.mbrc.events.ui.VolumeChangeEvent;
import com.kelsos.mbrc.interactors.MuteInteractor;
import com.kelsos.mbrc.interactors.PlayerInteractor;
import com.kelsos.mbrc.interactors.PlayerStateInteractor;
import com.kelsos.mbrc.interactors.RepeatInteractor;
import com.kelsos.mbrc.interactors.ShuffleInteractor;
import com.kelsos.mbrc.interactors.TrackCoverInteractor;
import com.kelsos.mbrc.interactors.TrackInfoInteractor;
import com.kelsos.mbrc.interactors.TrackPositionInteractor;
import com.kelsos.mbrc.interactors.VolumeInteractor;
import com.kelsos.mbrc.ui.views.MainView;
import com.kelsos.mbrc.utilities.ErrorHandler;
import com.kelsos.mbrc.utilities.RxBus;
import com.kelsos.mbrc.viewmodels.MainViewModel;
import java.util.concurrent.TimeUnit;
import roboguice.inject.ContextSingleton;
import rx.Observable;
import rx.Subscription;
import rx.android.schedulers.AndroidSchedulers;
import rx.schedulers.Schedulers;
import timber.log.Timber;

@ContextSingleton public class MainViewPresenterImpl implements MainViewPresenter {
  @Inject private ErrorHandler errorHandler;
  @Inject private MainViewModel model;
  @Inject private PlayerInteractor playerInteractor;
  @Inject private TrackInfoInteractor trackInfoInteractor;
  @Inject private VolumeInteractor volumeInteractor;
  @Inject private ShuffleInteractor shuffleInteractor;
  @Inject private RepeatInteractor repeatInteractor;
  @Inject private MuteInteractor muteInteractor;
  @Inject private TrackCoverInteractor coverInteractor;
  @Inject private TrackPositionInteractor positionInteractor;
  @Inject private PlayerStateInteractor playerStateInteractor;

  @Inject private RxBus bus;

  private MainView mainView;
  private Subscription positionUpdate;

  public MainViewPresenterImpl() {
  }

  @Override public void bind(MainView mainView) {
    this.mainView = mainView;
  }

  @Override public void onPause() {
    bus.unregister(this);
  }

  @Override public void onResume() {
    loadTrackInfo();
    loadCover();
    loadShuffle();
    loadRepeat();
    loadVolume();
    loadPosition();
    loadPlayerState();
    model.loadComplete();
    subscribe();
  }

  private void subscribe() {
    bus.registerOnMain(this, VolumeChangeEvent.class, this::onVolumeChangedEvent);
    bus.registerOnMain(this, RepeatChange.class, this::onRepeatChangedEvent);
    bus.registerOnMain(this, TrackInfoChangeEvent.class, this::onTrackInfoChangedEvent);
    bus.registerOnMain(this, CoverChangedEvent.class, this::onCoverChangedEvent);
    bus.registerOnMain(this, PlayStateChange.class, this::onPlayStateChanged);
    bus.registerOnMain(this, MuteChangeEvent.class, this::onMuteChanged);
  }

  private void loadPlayerState() {
    if (model.isLoaded()) {
      mainView.updatePlayState(model.getPlayState());
    } else {
      playerStateInteractor.getState().doOnNext(model::setPlayState).subscribe(state -> {
        mainView.updatePlayState(state);
        updatePlaystate(state);
      }, errorHandler::handleThrowable);
    }
  }

  private void loadTrackInfo() {
    if (model.getTrackInfo() == null) {
      trackInfoInteractor.execute(false).subscribe(trackInfo -> {
        model.setTrackInfo(trackInfo);
        mainView.updateTrackInfo(trackInfo);
      }, errorHandler::handleThrowable);
    } else {
      mainView.updateTrackInfo(model.getTrackInfo());
    }
  }

  private void loadCover() {
    if (model.getTrackCover() == null) {
      coverInteractor.execute(false).subscribe(bitmap -> {
        model.setTrackCover(bitmap);
        mainView.updateCover(bitmap);
      }, errorHandler::handleThrowable);
    } else {
      mainView.updateCover(model.getTrackCover());
    }
  }

  private void loadShuffle() {
    if (model.isLoaded()) {
      mainView.updateShuffle(model.getShuffle());
    } else {
      shuffleInteractor.getShuffle().subscribe(shuffle -> {
        model.setShuffle(shuffle);
        mainView.updateShuffle(shuffle);
      }, errorHandler::handleThrowable);
    }
  }

  private void loadRepeat() {
    if (model.isLoaded()) {
      mainView.updateRepeat(model.getRepeat());
    } else {
      repeatInteractor.getRepeat().doOnNext(model::setRepeat).subscribe(repeat -> {
        mainView.updateRepeat(repeat);
      }, errorHandler::handleThrowable);
    }
  }

  private void loadVolume() {
    if (model.isLoaded()) {
      mainView.updateVolume(model.getVolume());
    } else {
      volumeInteractor.getVolume().doOnNext(model::setVolume).subscribe(volume -> {
        mainView.updateVolume(volume);
      }, errorHandler::handleThrowable);
    }
  }

  private void loadPosition() {
    if (model.isLoaded()) {
      mainView.updatePosition(model.getPosition());
    } else {
      positionInteractor.getPosition().doOnNext(model::setPosition).subscribe(position -> {
        mainView.updatePosition(position);
      }, errorHandler::handleThrowable);
    }
  }

  @Override public void onPlayPausePressed() {
    performAction(PlayerAction.PLAY_PLAUSE);
  }

  @Override public void onPreviousPressed() {
    performAction(PlayerAction.PREVIOUS);
  }

  @Override public void onNextPressed() {
    performAction(PlayerAction.NEXT);
  }

  private void performAction(String action) {
    playerInteractor.performAction(action)
        .subscribeOn(Schedulers.io())
        .observeOn(AndroidSchedulers.mainThread())
        .subscribe(response -> {

        }, errorHandler::handleThrowable);
  }

  @Override public void onStopPressed() {
    performAction(PlayerAction.STOP);
  }

  @Override public void onMutePressed() {
    muteInteractor.toggle().subscribeOn(Schedulers.io()).observeOn(AndroidSchedulers.mainThread()).subscribe(isMute -> {
      model.setMuted(isMute);
      mainView.updateMute(isMute);
    }, errorHandler::handleThrowable);
  }

  @Override public void onShufflePressed() {
    shuffleInteractor.updateShuffle(Shuffle.TOGGLE).subscribe(shuffle -> {
      model.setShuffle(shuffle);
      mainView.updateShuffle(shuffle);
    }, errorHandler::handleThrowable);
  }

  @Override public void onRepeatPressed() {
    repeatInteractor.setRepeat(Repeat.CHANGE)
        .subscribeOn(Schedulers.io())
        .subscribeOn(Schedulers.io())
        .observeOn(AndroidSchedulers.mainThread())
        .subscribe(mainView::updateRepeat, errorHandler::handleThrowable);
  }

  @Override public void onVolumeChange(int volume) {
    volumeInteractor.setVolume(volume)
        .subscribeOn(Schedulers.io())
        .observeOn(AndroidSchedulers.mainThread())
        .subscribe(model::setVolume, errorHandler::handleThrowable);
  }

  @Override public void onPositionChange(int position) {
    stopPositionUpdate();
    updatePosition(positionInteractor.setPosition(position));
  }

  public void updatePosition(Observable<TrackPosition> positionObservable) {
    positionObservable.subscribeOn(Schedulers.io())
        .observeOn(AndroidSchedulers.mainThread())
        .doOnNext(model::setPosition)
        .subscribe((newPosition) -> {
          startPositionUpdate();
        }, errorHandler::handleThrowable);
  }

  @Override public void onScrobbleToggle() {

  }

  @Override public void onLfmLoveToggle() {

  }

  public void onVolumeChangedEvent(VolumeChangeEvent event) {
    model.setVolume(event.getVolume());
    mainView.updateVolume(event.getVolume());
  }

  public void onRepeatChangedEvent(RepeatChange event) {
    model.setRepeat(event.getMode());
    mainView.updateRepeat(event.getMode());
  }

  public void onTrackInfoChangedEvent(TrackInfoChangeEvent event) {
    Timber.v("Received change event");
    model.setTrackInfo(event.getTrackInfo());
    mainView.updateTrackInfo(event.getTrackInfo());
    startPositionUpdate();
    updatePosition(positionInteractor.getPosition());
  }

  public void onCoverChangedEvent(CoverChangedEvent event) {
    model.setTrackCover(event.getCover());
    mainView.updateCover(event.getCover());
  }

  public void onPlayStateChanged(PlayStateChange event) {
    model.setPlayState(event.getState());
    mainView.updatePlayState(event.getState());
    updatePlaystate(event.getState());
  }

  private void updatePlaystate(String state) {
    if (PlayerState.PLAYING.equals(state)) {
      startPositionUpdate();
    } else {
      stopPositionUpdate();
    }
  }

  private void stopPositionUpdate() {
    Timber.v("Track now is either paused or stoped");
    if (positionUpdate != null && !positionUpdate.isUnsubscribed()) {
      positionUpdate.unsubscribe();
      positionUpdate = null;
    }
  }

  private void startPositionUpdate() {
    if (positionUpdate != null) {
      return;
    }

    Timber.v("Track is now playing");
    positionUpdate = Observable.interval(0, 1, TimeUnit.SECONDS)
        .subscribeOn(Schedulers.io())
        .observeOn(AndroidSchedulers.mainThread())
        .takeWhile(aLong1 -> {
          final TrackPosition position = model.getPosition();
          return position.getCurrent() < position.getTotal();
        })
        .subscribe(aLong -> {
          final TrackPosition modelPosition = model.getPosition();
          TrackPosition newPosition = new TrackPosition(modelPosition.getCurrent() + 1000, modelPosition.getTotal());
          model.setPosition(newPosition);
          mainView.updatePosition(newPosition);
        }, errorHandler::handleThrowable);
  }

  public void onMuteChanged(MuteChangeEvent event) {
    model.setMuted(event.isMute());
    mainView.updateMute(event.isMute());
  }
}

File: app/src/main/kotlin/com/kelsos/mbrc/presenters/MainViewPresenterImpl.kt
package com.kelsos.mbrc.presenters

import com.kelsos.mbrc.annotations.PlayerAction
import com.kelsos.mbrc.annotations.PlayerState
import com.kelsos.mbrc.annotations.Repeat
import com.kelsos.mbrc.annotations.Shuffle
import com.kelsos.mbrc.domain.TrackPosition
import com.kelsos.mbrc.events.ui.CoverChangedEvent
import com.kelsos.mbrc.events.ui.MuteChangeEvent
import com.kelsos.mbrc.events.ui.PlayStateChange
import com.kelsos.mbrc.events.ui.RepeatChange
import com.kelsos.mbrc.events.ui.TrackInfoChangeEvent
import com.kelsos.mbrc.events.ui.VolumeChangeEvent
import com.kelsos.mbrc.extensions.task
import com.kelsos.mbrc.interactors.MuteInteractor
import com.kelsos.mbrc.interactors.PlayerInteractor
import com.kelsos.mbrc.interactors.PlayerStateInteractor
import com.kelsos.mbrc.interactors.RepeatInteractor
import com.kelsos.mbrc.interactors.ShuffleInteractor
import com.kelsos.mbrc.interactors.TrackCoverInteractor
import com.kelsos.mbrc.interactors.TrackInfoInteractor
import com.kelsos.mbrc.interactors.TrackPositionInteractor
import com.kelsos.mbrc.interactors.VolumeInteractor
import com.kelsos.mbrc.ui.views.MainView
import com.kelsos.mbrc.utilities.ErrorHandler
import com.kelsos.mbrc.utilities.RxBus
import com.kelsos.mbrc.viewmodels.MainViewModel
import rx.Observable
import rx.Subscription
import rx.subjects.PublishSubject
import timber.log.Timber
import toothpick.smoothie.annotations.ContextSingleton
import java.util.concurrent.TimeUnit
import javax.inject.Inject

@ContextSingleton class MainViewPresenterImpl : MainViewPresenter {
  override fun attachView(view: MainView) {
    this.mainView = view

    loadTrackInfo()
    loadCover()
    loadShuffle()
    loadRepeat()
    loadVolume()
    loadPosition()
    loadPlayerState()
    model.loadComplete()
    subscribe()
  }

  override fun detachView() {
    this.mainView = null
    bus.unregister(this)
  }

  @Inject private lateinit var errorHandler: ErrorHandler
  @Inject private lateinit var model: MainViewModel
  @Inject private lateinit var playerInteractor: PlayerInteractor
  @Inject private lateinit var trackInfoInteractor: TrackInfoInteractor
  @Inject private lateinit var volumeInteractor: VolumeInteractor
  @Inject private lateinit var shuffleInteractor: ShuffleInteractor
  @Inject private lateinit var repeatInteractor: RepeatInteractor
  @Inject private lateinit var muteInteractor: MuteInteractor
  @Inject private lateinit var coverInteractor: TrackCoverInteractor
  @Inject private lateinit var positionInteractor: TrackPositionInteractor
  @Inject private lateinit var playerStateInteractor: PlayerStateInteractor

  @Inject private lateinit var bus: RxBus

  private var mainView: MainView? = null
  private var positionUpdate: Subscription? = null
  private var volumeThrottler: PublishSubject<Int> = PublishSubject.create()

  init {
    volumeThrottler.throttleLast(500, TimeUnit.MILLISECONDS)
        .subscribe { changeVolume(it) }
  }

  private fun subscribe() {
    bus.registerOnMain(this, VolumeChangeEvent::class.java, { this.onVolumeChangedEvent(it) })
    bus.registerOnMain(this, RepeatChange::class.java, { this.onRepeatChangedEvent(it) })
    bus.registerOnMain(this, TrackInfoChangeEvent::class.java, { this.onTrackInfoChangedEvent(it) })
    bus.registerOnMain(this, CoverChangedEvent::class.java, { this.onCoverChangedEvent(it) })
    bus.registerOnMain(this, PlayStateChange::class.java, { this.onPlayStateChanged(it) })
    bus.registerOnMain(this, MuteChangeEvent::class.java, { this.onMuteChanged(it) })
  }

  private fun loadPlayerState() {
    if (model.isLoaded) {
      mainView?.updatePlayState(model.playState)
    } else {
      playerStateInteractor.getState().doOnNext({ model.playState = it }).subscribe({
        mainView?.updatePlayState(it)
        updatePlaystate(it)
      }, { errorHandler.handleThrowable(it) })
    }
  }

  private fun loadTrackInfo() {
    if (model.isLoaded) {
      mainView?.updateTrackInfo(model.trackInfo)
    } else {
      trackInfoInteractor.load().task().subscribe({
        model.trackInfo = it
        mainView?.updateTrackInfo(it)
      }, { errorHandler.handleThrowable(it) })
    }
  }

  private fun loadCover() {
    if (model.isLoaded) {
      mainView?.updateCover(model.trackCover)
    } else {
      coverInteractor.load().subscribe({
        model.trackCover = it
        mainView?.updateCover(it)
      }, { errorHandler.handleThrowable(it) })
    }
  }

  private fun loadShuffle() {
    if (model.isLoaded) {
      mainView?.updateShuffle(model.shuffle)
    } else {
      shuffleInteractor.getShuffle().subscribe({
        model.shuffle = it
        mainView?.updateShuffle(it)
      }, { errorHandler.handleThrowable(it) })
    }
  }

  private fun loadRepeat() {
    if (model.isLoaded) {
      mainView?.updateRepeat(model.repeat)
    } else {
      repeatInteractor.getRepeat()
          .doOnNext({ model.repeat = it })
          .subscribe({
            mainView?.updateRepeat(it)
          }, { errorHandler.handleThrowable(it) })
    }
  }

  private fun loadVolume() {
    if (model.isLoaded) {
      mainView?.updateVolume(model.volume)
    } else {
      volumeInteractor.getVolume().doOnNext({ model.volume = it })
          .subscribe({
            mainView?.updateVolume(it!!)
          }, { errorHandler.handleThrowable(it) })
    }
  }

  private fun loadPosition() {
    if (model.isLoaded) {
      mainView?.updatePosition(model.position)
    } else {
      positionInteractor.getPosition()
          .doOnNext {
            model.position = it
          }
          .subscribe({ mainView?.updatePosition(it) }) { errorHandler.handleThrowable(it) }
    }
  }

  override fun onPlayPausePressed() {
    performAction(PlayerAction.PLAY_PLAUSE)
  }

  override fun onPreviousPressed() {
    performAction(PlayerAction.PREVIOUS)
  }

  override fun onNextPressed() {
    performAction(PlayerAction.NEXT)
  }

  private fun performAction(action: String) {
    playerInteractor.performAction(action).task().subscribe(
        {

        }
    ) { errorHandler.handleThrowable(it) }
  }

  override fun onStopPressed() {
    performAction(PlayerAction.STOP)
  }

  override fun onMutePressed() {
    muteInteractor.toggle().task().subscribe({
      model.isMuted = it!!
      mainView?.updateMute(it)
    }) { errorHandler.handleThrowable(it) }
  }

  override fun onShufflePressed() {
    shuffleInteractor.updateShuffle(Shuffle.TOGGLE).subscribe({
      model.shuffle = it
      mainView?.updateShuffle(it)
    }) {
      errorHandler.handleThrowable(it)
    }
  }

  override fun onRepeatPressed() {
    repeatInteractor.setRepeat(Repeat.CHANGE).task()
        .doOnNext {
          model.repeat = it
        }
        .subscribe({
          mainView?.updateRepeat(it)
        })
        {
          errorHandler.handleThrowable(it)
        }
  }

  override fun onVolumeChange(volume: Int) {
    volumeThrottler.onNext(volume)
  }

  private fun changeVolume(volume: Int) {
    volumeInteractor.setVolume(volume)
        .task()
        .subscribe({ model.volume = it }) {
          errorHandler.handleThrowable(it)
        }
  }

  override fun onPositionChange(position: Int) {
    stopPositionUpdate()
    updatePosition(positionInteractor.setPosition(position))
  }

  fun updatePosition(positionObservable: Observable<TrackPosition>) {
    positionObservable.task().doOnNext({ model.position = it })
        .subscribe({ startPositionUpdate() }
        ) { errorHandler.handleThrowable(it) }
  }

  override fun onScrobbleToggle() {

  }

  override fun onLfmLoveToggle() {

  }

  fun onVolumeChangedEvent(event: VolumeChangeEvent) {
    model.volume = event.volume
    mainView?.updateVolume(event.volume)
  }

  fun onRepeatChangedEvent(event: RepeatChange) {
    model.repeat = event.mode
    mainView?.updateRepeat(event.mode)
  }

  fun onTrackInfoChangedEvent(event: TrackInfoChangeEvent) {
    Timber.v("Received change event")
    model.trackInfo = event.trackInfo
    mainView?.updateTrackInfo(event.trackInfo)
    startPositionUpdate()
    updatePosition(positionInteractor.getPosition())
  }

  fun onCoverChangedEvent(event: CoverChangedEvent) {
    model.trackCover = event.cover
    mainView?.updateCover(event.cover)
  }

  fun onPlayStateChanged(event: PlayStateChange) {
    model.playState = event.state
    mainView?.updatePlayState(event.state)
    updatePlaystate(event.state)
  }

  private fun updatePlaystate(state: String) {
    Timber.v("New State $state")
    if (PlayerState.PLAYING == state) {
      loadPosition()
      startPositionUpdate()
    } else {
      stopPositionUpdate()
    }
  }

  private fun stopPositionUpdate() {
    Timber.v("Track is not playing")
    positionUpdate?.unsubscribe()
    positionUpdate = null
  }

  private fun startPositionUpdate() {
    if (positionUpdate != null) {
      return
    }

    Timber.v("Track is playing")
    positionUpdate = Observable.interval(0, 1, TimeUnit.SECONDS).task().takeWhile {
      val position = model.position
      position.current < position.total
    }.subscribe({
      val modelPosition = model.position
      val newPosition = TrackPosition(modelPosition.current + 1000, modelPosition.total)
      model.position = newPosition
      mainView?.updatePosition(newPosition)
    }, { errorHandler.handleThrowable(it) })
  }

  fun onMuteChanged(event: MuteChangeEvent) {
    model.isMuted = event.isMute
    mainView?.updateMute(event.isMute)
  }
}
